### ISSUE TITLE
Terraform crashes when I run a tf file which contain dynamic

### ISSUE THREAD

**Issue by Appledelta** [hashicorp/terraform#24838]
I am using Terraform verson 0.12.24 with AWS provider version 2.60

I am running the following tf file and terraform crashes. I have attached the crash.log file.
[crash.log](https://github.com/hashicorp/terraform/files/4571487/crash.log)
[crash.log](https://github.com/hashicorp/terraform/files/4571488/crash.log)

```HCL
variable "sg_ports" {
  type          = list(number)
  description = "List of ingress ports"
  default     = [8200, 8201, 8300, 9200, 9500]
}

resource "aws_security_groups" "dynamicsg" {
  name        = "dynamicsg"
  description = "Ingress for Vault"
  
dynamic "ingress" {
  for_each = var.sg_ports
  content {
    from_port   = ingress.value
    to_port     = ingress.value
    protocol = "tcp"
    cidr_blocks = ["0.0.0.0/0"]
  }
}
}
```

**Comment by danieldreier** [comment]
@Appledelta thank you so much for reporting this! I was able to [reproduce it](https://github.com/danieldreier/terraform-issue-reproductions/tree/master/24838).

**Comment by edgarsandi** [comment]
Hi all!

We don't have the resource called `aws_security_groups` we have a data source with this name.
[Data Source: aws_security_groups](https://www.terraform.io/docs/providers/aws/d/security_groups.html)
[Resource: aws_security_group](https://www.terraform.io/docs/providers/aws/r/security_group.html)

@danieldreier @Appledelta  If you change to use the resource, will work correctly
``` git
@@ -9,7 +9,7 @@ variable "sg_ports" {
   default     = [8200, 8201, 8300, 9200, 9500]
 }

-resource "aws_security_groups" "dynamicsg" {
+resource "aws_security_group" "dynamicsg" {
   name        = "dynamicsg"
   description = "Ingress for Vault"

(END)
```

``` hcl
$> terraform --version
Terraform v0.12.26
+ provider.aws v2.64.0
```

``` hcl
$> terraform init
Initializing the backend...

Initializing provider plugins...

Terraform has been successfully initialized!

You may now begin working with Terraform. Try running "terraform plan" to see
any changes that are required for your infrastructure. All Terraform commands
should now work.

If you ever set or change modules or backend configuration for Terraform,
rerun this command to reinitialize your working directory. If you forget, other
commands will detect it and remind you to do so if necessary.
```

``` hcl
$> terraform validate
Success! The configuration is valid.
```

[]'s

**Comment by mildwonkey** [comment]
HI all! I just merged a fix for this bug which will be included in terraform v0.13. Thank you for reporting it.
